### PR TITLE
Tril indices

### PR DESCRIPTION
**Pull request by zhumakhan**
#6334

### CODE REVIEW

**Comment by zhumakhan** [comment]
+minor fixes for triu_indices
+tril_indices extension

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

File: ivy/container/extensions/creation.py [hunk 1]
```diff
@@ -31,11 +30,10 @@ def static_triu_indices(
             prune_unapplied,
             map_sequences,
             device=device,
-            out=out,
         )
 
-    def triu_indices(
-        self: ivy.Container,
```
Comment (CatB1t) [inline ivy/container/extensions/creation.py:38]: We shouldn't be removing other code that isn't related to the PR, could you go over the changes and make sure no unrelated changes are made?

File: ivy_tests/test_ivy/test_functional/test_extensions/test_core/test_creation.py [hunk 2]
```diff
@@ -36,3 +37,36 @@ def test_triu_indices(
         k=k,
         device=device,
     )
+
+
+@handle_cmd_line_args
```
Comment (CatB1t) [inline ivy_tests/test_ivy/test_functional/test_extensions/test_core/test_creation.py:42]: We should be using the new decorator `handle_test`, look here for more info: https://lets-unify.ai/ivy/deep_dive/ivy_tests.html#ivy-test-decorators

File: ivy/functional/backends/jax/extensions/creation.py [hunk 3]
```diff
@@ -20,8 +20,22 @@ def triu_indices(
     /,
     *,
     device: jaxlib.xla_extension.Device,
-) -> Tuple[JaxArray]:
+) -> Tuple[JaxArray, ...]:
```
Comment (CatB1t) [inline ivy/functional/backends/jax/extensions/creation.py:23]: I'm not sure what `...` should be indicating in the type hinting here, can you elaborate more why we should be adding `...` to return type hint?

File: ivy/container/extensions/creation.py (hunk 1 above)
Comment (zhumakhan) [reply ivy/container/extensions/creation.py]: Got it, I will create separate PR for it.

File: ivy_tests/test_ivy/test_functional/test_extensions/test_core/test_creation.py (hunk 2 above)
Comment (zhumakhan) [reply ivy_tests/test_ivy/test_functional/test_extensions/test_core/test_creation.py]: Done.

File: ivy/functional/backends/jax/extensions/creation.py (hunk 3 above)
Comment (zhumakhan) [reply ivy/functional/backends/jax/extensions/creation.py]: So Tuple[<type>] is a hint for a tuple with a single element of type <type>,
Tupe[<type>, ...] is hint for a tuple with one or multiple elements with type <type>

**Comment by ivy-seed** [comment]
This PR has been closed because it has been marked as stale for more than 7 days with no activity.

**Comment by zhumakhan** [comment]
Hi @CatB1t, thanks a lot for your comments, requested changes are made

File: ivy/functional/experimental/creation.py [hunk 4]
```diff
@@ -384,3 +384,104 @@ def hamming_window(
...
+@handle_exceptions
+def tril_indices(
+    n_rows: int,
+    n_cols: Optional[int] = None,
+    k: Optional[int] = 0,
+    /,
+    *,
+    device: Optional[Union[ivy.Device, ivy.NativeDevice]] = None,
+) -> Tuple[ivy.Array, ...]:
+    """Returns the indices of the upper triangular part of a row by col matrix in a
```
Comment (CatB1t) [inline ivy/functional/experimental/creation.py:401]: Lower triangular, not upper.

File: ivy/functional/experimental/creation.py [hunk 5]
```diff
@@ -384,3 +384,104 @@ def hamming_window(
...
+    2-by-N shape (tuple of two N dimensional arrays), where the first row contains
+    row coordinates of all indices and the second row contains column coordinates.
+    Indices are ordered based on rows and then columns.  The upper triangular part
```
Comment (CatB1t) [inline ivy/functional/experimental/creation.py:404]: Lower triangular, not upper.

File: ivy/functional/experimental/creation.py [hunk 6]
```diff
@@ -384,3 +384,104 @@ def hamming_window(
...
+    k controls which diagonal to consider. If k = 0, all elements on and above the main
+    diagonal are retained. A positive value excludes just as many diagonals above the
+    main diagonal, and similarly a negative value includes just as many diagonals
+    below the main diagonal. The main diagonal are the set of indices
+    {(i,i)} for i∈[0,min{n_rows, n_cols}−1].
+
+    Notes
+    -----
+    Primary purpose of this function is to slice an array of shape (n,m). See
+    https://numpy.org/doc/stable/reference/generated/numpy.triu_indices.html
```
Comment (CatB1t) [inline ivy/functional/experimental/creation.py:415]: Link mentions wrong function.

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_creation.py [hunk 7]
```diff
@@ -277,3 +277,44 @@ def test_hamming_window(
...
+    ground_truth_backend,
+):
+    helpers.test_function(
+        ground_truth_backend=ground_truth_backend,
+        input_dtypes=["int32"],
+        as_variable_flags=[False],
+        with_out=False,
+        num_positional_args=num_positional_args,
+        native_array_flags=native_array,
+        container_flags=[False],
```
Comment (CatB1t) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_creation.py:311]: We shouldn't hardcode this, instead use `container_flags`.

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_creation.py [hunk 8]
```diff
@@ -277,3 +277,44 @@ def test_hamming_window(
...
+    instance_method,
+    backend_fw,
+    fn_name,
+    on_device,
+    ground_truth_backend,
+):
+    helpers.test_function(
+        ground_truth_backend=ground_truth_backend,
+        input_dtypes=["int32"],
+        as_variable_flags=[False],
```
Comment (CatB1t) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_creation.py:307]: We shouldn't hardcode this, instead use `as_variable` flag.

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_creation.py [hunk 9]
```diff
@@ -277,3 +277,44 @@ def test_hamming_window(
...
+    container_flags,
+    instance_method,
+    backend_fw,
+    fn_name,
+    on_device,
+    ground_truth_backend,
+):
+    helpers.test_function(
+        ground_truth_backend=ground_truth_backend,
+        input_dtypes=["int32"],
```
Comment (CatB1t) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_creation.py:306]: This should be randomally generated.

**Review by CatB1t — CHANGES_REQUESTED** [review]
Hello @zhumakhan , Great work! just few points to address, please read them carefully, also torch seems to be failing the test, here an example that makes `torch` backend diverge from `tensorflow` backend: 

```python
ivy.set_backend("tensorflow")
print(ivy.functional.experimental.tril_indices(0, 2, 2))
ivy.set_backend("torch")
print(ivy.functional.experimental.tril_indices(0, 2, 2))
```
could you please fix that and re-request a review from me? Thanks!

**Comment by zhumakhan** [comment]
Thanks for comments and mentioning error with pytorch backend. Requested changes are made, errors are fixed.

**Review by CatB1t — CHANGES_REQUESTED** [review]
Hello @zhumakhan , thanks for the updates! `torch` backend is failing due to another error: `torch: tril_indices: tril_indices(): argument 'col' must be int, not NoneType`, could you please fix that and request a review again? Thanks!

**Comment by zhumakhan** [comment]
Thanks again @CatB1t , mentioned issues are fixed

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_creation.py [hunk 10]
```diff
@@ -277,3 +277,47 @@ def test_hamming_window(
...
         dtype=dtype,
     )
+
+
+@handle_test(
+    fn_tree="functional.experimental.tril_indices",
+    n_rows=helpers.ints(min_value=0, max_value=10),
+    n_cols=st.none() | helpers.ints(min_value=0, max_value=10),
+    k=helpers.ints(min_value=-11, max_value=11),
+    dtypes_x=helpers.dtype_and_values(available_dtypes=("int32", "int64")),
```
Comment (CatB1t) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_creation.py:287]: To make this more efficient, we should directly use `helpers.get_dtypes()` with `full=False`, so we do not generate an array.

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_creation.py [hunk 11]
```diff
@@ -277,3 +277,47 @@ def test_hamming_window(
...
+)
+def test_tril_indices(
+    *,
+    n_rows,
+    n_cols,
+    k,
+    num_positional_args,
+    as_variable,
+    # with_out,
```
Comment (CatB1t) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_creation.py:296]: We should remove the parameter and not just comment it out.

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_creation.py [hunk 12]
```diff
@@ -277,3 +277,47 @@ def test_hamming_window(
...
+    dtypes, _ = dtypes_x
+    helpers.test_function(
+        input_dtypes=dtypes,
+        ground_truth_backend=ground_truth_backend,
+        as_variable_flags=as_variable,
+        with_out=False,
+        num_positional_args=num_positional_args,
+        native_array_flags=native_array,
+        container_flags=container_flags,
+        instance_method=instance_method,
```
Comment (CatB1t) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_creation.py:315]: The test are failing due to generated as `instance_method=True` which means we should test the function as an instance method, which is something we shouldn't do here, as there's no instance method of `tril_indicies`, we should change this to `instance_method=False`.

**Review by CatB1t — CHANGES_REQUESTED** [review]

**Comment by zhumakhan** [comment]
Tests are fixed now, Could you please review?
Thanks

**Review by CatB1t — APPROVED** [review]
lgtm! thanks for your contribution!
